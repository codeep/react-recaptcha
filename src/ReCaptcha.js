import React, { Component } from "react";
import PropTypes from "prop-types";

const isReady = () =>
  typeof window !== "undefined" &&
  window.grecaptcha !== undefined &&
  // need to check for presence of render, because reCaptcha creates { ready } first
  !!window.grecaptcha.render;

const defaultProps = {
  theme: "light",
  size: "normal",
  tabIndex: 0,
  badge: "bottomright",
  isolated: true,
  inherit: false
};

const propTypes = {
  sitekey: PropTypes.string.isRequired,
  /** Optional. The callback function, executed when the user submits a successful response. The g-recaptcha-response token is passed to your callback. */
  onSuccess: PropTypes.func,
  /** Optional. The callback function to be executed once all the dependencies have loaded. */
  onLoad: PropTypes.func,
  /** Optional. The callback function, executed when the reCAPTCHA response expires and the user needs to re-verify. */
  onExpired: PropTypes.func,
  /** Optional. The callback function, executed when reCAPTCHA encounters an error (usually network connectivity) and cannot continue until connectivity is restored. If you specify a function here, you are responsible for informing the user that they should retry. */
  onError: PropTypes.func,
  /** Optional. The tabIndex of the challenge. If other elements in your page use tabIndex, it should be set to make user navigation easier. */
  tabIndex: PropTypes.number,
  /** "invisible" - stands for Invisible reCAPTCHA
   *  "compact", "normal" - stands for I'm not a robot challenge */
  size: PropTypes.oneOf(["compact", "normal", "invisible"]),
  /** Works with I'm not a robot challenge
   *  Optional. The color theme of the widget. */
  theme: PropTypes.oneOf(["dark", "light"]),
  /** Works with Invisible reCAPTCHA
   *  Optional. Reposition the reCAPTCHA badge. 'inline' lets you position it with CSS. */
  badge: PropTypes.oneOf(["bottomright", "bottomleft", "inline"]),
  /** Works with Invisible reCAPTCHA
   *  Optional. For plugin owners to not interfere with existing reCAPTCHA installations on a page. If true, this reCAPTCHA instance will be part of a separate ID space. */
  isolated: PropTypes.bool,
  /** Use existing data-* attributes on the element if the coorsponding parameter is not specified. The values in parameter will take precedence over the attributes. */
  inherit: PropTypes.bool
};

class ReCaptcha extends Component {
  constructor(props) {
    super(props);
    this.recaptcha = null;
    this._setRecaptcha = this._setRecaptcha.bind(this);
    this.widget = null;
    this.reset = this.reset.bind(this);
    this.execute = this.execute.bind(this);
    this._updateReadyState = this._updateReadyState.bind(this);
    this._renderGrecaptcha = this._renderGrecaptcha.bind(this);
  }

  componentDidMount() {
    if (!isReady()) {
      this.readyCheck = setInterval(this._updateReadyState, 500);
    }
  }

  componentWillUnmount() {
    clearInterval(this.readyCheck);
  }

  _setRecaptcha(recaptcha) {
    this.recaptcha = recaptcha;
    if (isReady() && this.recaptcha) {
      this._renderGrecaptcha();
    }
  }

  _updateReadyState() {
    if (isReady()) {
      clearInterval(this.readyCheck);
      if (this.recaptcha) this._renderGrecaptcha();
    }
  }

  _renderGrecaptcha() {
    const {
      sitekey,
      onSuccess,
      theme,
      type,
      size,
      tabIndex,
      badge,
      onError,
      onExpired,
      inherit
    } = this.props;
    this.widget = window.grecaptcha.render(
      this.recaptcha,
      {
        sitekey,
        callback: onSuccess,
        theme,
        type,
        size,
        tabindex: tabIndex,
        badge,
        "error-callback": onError,
        "expired-callback": onExpired
      },
      inherit
    );
    const { onLoad } = this.props;
    if (onLoad) onLoad();
  }

  reset() {
    if (isReady() && this.widget !== null) window.grecaptcha.reset(this.widget);
  }

  execute() {
    if (this.props.size !== "invisible") {
      if (process.env.NODE_ENV !== "production") {
        throw new Error(
          'You need to use `execute` only for `size="invisible"`'
        );
      }
      // otherwise user will se confusing error: A cross-origin error was thrown
      return;
    }
    if (isReady() && this.widget !== null)
      window.grecaptcha.execute(this.widget);
  }

  render() {
    const {
      sitekey,
      onSuccess,
      onLoad,
      onExpired,
      onError,
      tabIndex,
      size,
      theme,
      badge,
      isolated,
      inherit
    } = this.props;
    return (
      <div
        ref={this._setRecaptcha}
        data-sitekey={sitekey}
        data-theme={theme}
        data-size={size}
        data-badge={badge}
        data-tabindex={`${tabIndex}`}
      />
    );
  }
}

ReCaptcha.defaultProps = defaultProps;
ReCaptcha.propTypes = propTypes;

export default ReCaptcha;
