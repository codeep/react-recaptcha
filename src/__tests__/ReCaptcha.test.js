import React from "react";
import renderer from "react-test-renderer";
import ReCaptcha from "../ReCaptcha";

const pause = time => new Promise(resolve => setTimeout(resolve, time));

const mockGlobal = grecaptcha => {
  window.grecaptcha = {
    render: jest.fn(),
    ...grecaptcha
  };
};
const resetMock = () => {
  window.grecaptcha = undefined;
};

const subject = (options = {}) =>
  renderer.create(<ReCaptcha sitekey="test" {...options} />, {
    createNodeMock: element => ({ type: `mock-${element.type}` })
  });

describe("ReCaptcha", () => {
  afterEach(resetMock);

  test("render", () => {
    const component = subject();
    expect(component.toJSON()).toMatchSnapshot();
    component.unmount();
  });

  test("onLoad, preloaded", () => {
    const onLoad = jest.fn();
    const render = jest.fn();
    mockGlobal({ render });
    const component = subject({ onLoad });
    // it calls onLoad as soon as it mounted
    expect(onLoad.mock.calls.length).toEqual(1);
    // it calls grecaptcha.render
    expect(render.mock.calls.length).toEqual(1);
    expect(render.mock.calls).toMatchSnapshot();
    component.unmount();
  });

  test("onLoad, initially not loaded", async () => {
    const onLoad = jest.fn();
    const component = subject({ onLoad });
    expect(onLoad.mock.calls.length).toEqual(0);
    const render = jest.fn();
    mockGlobal({ render });
    // it waits for grecaptcha to load
    await pause(510);
    expect(onLoad.mock.calls.length).toEqual(1);
    // it calls grecaptcha.render
    expect(render.mock.calls.length).toEqual(1);
    expect(render.mock.calls).toMatchSnapshot();
    component.unmount();
  });

  test("reset", () => {
    const reset = jest.fn();
    mockGlobal({ reset });
    const component = subject();
    component.getInstance().reset();
    // it calls grecaptcha.reset
    expect(reset.mock.calls.length).toEqual(1);
    expect(reset.mock.calls).toMatchSnapshot();
    component.unmount();
  });

  test("execute for visible", () => {
    const execute = jest.fn();
    mockGlobal({ execute });
    const component = subject();
    try {
      component.getInstance().execute();
    } catch (e) {
      expect(e.message).toEqual(
        'You need to use `execute` only for `size="invisible"`'
      );
    }
    // it doesn't call grecaptcha.execute
    expect(execute.mock.calls.length).toEqual(0);
    component.unmount();
  });

  test("execute for invisible", () => {
    const execute = jest.fn();
    mockGlobal({ execute });
    const component = subject({ size: "invisible" });
    component.getInstance().execute();
    // it calls grecaptcha.execute
    expect(execute.mock.calls.length).toEqual(1);
    expect(execute.mock.calls).toMatchSnapshot();
    component.unmount();
  });
});
