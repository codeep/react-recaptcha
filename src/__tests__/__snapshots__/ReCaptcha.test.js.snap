// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReCaptcha execute for invisible 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReCaptcha onLoad, initially not loaded 1`] = `
Array [
  Array [
    Object {
      "type": "mock-div",
    },
    Object {
      "badge": "bottomright",
      "callback": undefined,
      "error-callback": undefined,
      "expired-callback": undefined,
      "sitekey": "test",
      "size": "normal",
      "tabindex": 0,
      "theme": "light",
      "type": undefined,
    },
    false,
  ],
]
`;

exports[`ReCaptcha onLoad, preloaded 1`] = `
Array [
  Array [
    Object {
      "type": "mock-div",
    },
    Object {
      "badge": "bottomright",
      "callback": undefined,
      "error-callback": undefined,
      "expired-callback": undefined,
      "sitekey": "test",
      "size": "normal",
      "tabindex": 0,
      "theme": "light",
      "type": undefined,
    },
    false,
  ],
]
`;

exports[`ReCaptcha render 1`] = `
<div
  data-badge="bottomright"
  data-sitekey="test"
  data-size="normal"
  data-tabindex="0"
  data-theme="light"
/>
`;

exports[`ReCaptcha reset 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;
